### ISSUE TITLE
Tagging

### ISSUE THREAD

**Issue by rzezeski** [basho/yokozuna#5]
Data stored in Riak is opaque to Riak.  This means that Riak doesn't
know anything about the structure of the data being stored in it.
Whether it be JSON or JPEG, it's all the same to Riak.

On the contrary, the application storing the data often understands
the structure of the data.  The application may want to _tag_ the data
with attributes that give additional context.  For example, tagging a
picture with information such as who uploaded it and when it was
taken.

The Riak Object contains metadata which is a set of key-value pairs.
This lends itself nicely to indexing because Solr expects a document
which is a set of field-value pairs.  It's a matter of mapping the
metadata key to a field name.

In Riak the object metadata is an Erlang dict which is created
directly in the case of protobuffs or translated from HTTP headers in
the case of HTTP.  This feature should work the same no matter the
method of creation.
## API and Implementation Notes

TODO: UTF-8 in field names?
- The object's metadata may contain a `yz-tags` key.  The value should
  be a CSV specifying the fields to tag.  A comma must be used as
  delimiter and whitespace is ignored.
- For HTTP the header `x-yz-tags` may be used.  Yokozuna will check
  for both fields.
- If a field listed in `yz-tags` doesn't exist it will not be
  considered an error but will be silently ignored.
- Tag names should match Solr field names declared in the schema.
- If a tag field name doesn't match any fields defined in the Solr
  schema then the result will depend on the schema config.  In most
  cases I imagine this will cause an error.
- The value of a tag is always passed verbatim to Solr.
- Any metadata key may be tagged.
## Example

The following is an example of HTTP headers conforming with the rules
above.  The data is a fictitious picture uploaded by `myapp`.

```
yz-tags: myapp-where_loc, myapp-user_s, myapp-description_t
myapp-where_loc: Baltimore
myapp-user_s: rzezeski
myapp-description_t: Federal Hill at dusk.
```

The schema for this index might contain the following entries.

```
<dynamicField name="*_loc" type="location" indexed="true" stored="true"/>
<dynamicField name="*_s" type="string" indexed="true" stored="true" />
<dynamicField name="*_t" type="text_general" indexed="true" stored="false"/>
```

This example uses dynamic fields for each tag but a direct mapping
could be used as well.

```
yz-tags: myapp-description
myapp-description: Federal Hill at dusk.

<dynamicField name="myapp-description" type="text_general" indexed="true" stored="false"/>
```

An example of including metadata not specific to myapp.

```
yz-tags: content-type
```
## Thought Process

Following is a summary of the thought process for this feature.
### Mapping Keys to Fields

The first thing to decide is how to map a metadata key to a Solr
field.  A naive approach might look like the 2i interface which relies
on a prefix and a suffix.  The prefix signals the the key should be
indexed, e.g. `x-riak-index`.  The suffix indicates the type of the
value, e.g. `_bin`.  It's then a matter of deciding what the prefix
should be and how to map suffix to Solr fields.

In fact, why not use the 2i interface since it is already defined and
supported by Riak clients?  There are several problems with this
approach.
- There is only support for binary (`_bin`) and int (`_int`) types.
  Solr supports many more types.
- What Solr type should `_bin` map to?  It could reasonably map to
  different types like `text`, `text_general`, `string`, etc.
- What if `_bin` should be mapped to a different type depending on the
  data?  What is the convention for specifying the Solr type?
- What happens when leveldb is being used?  Should both 2i and Solr
  indexing occur?  What if the user only wants one of the systems to
  index the data?  How does the user tell the system what to do?
- The 2i interface is now performing double duty.  Any change to its
  API needs to take both systems into account.

Yokozuna could have a dedicated prefix/suffix convention but then
there is an additional layer of translation between Yokozuna and Solr.
This adds complexity in the system and causes additional burden to the
users.

Solr already has a mapping from field name to type.  This should be
exploited.  If a user adds a `summary` tag then there should be a
`summary` field in the Solr schema.  If the user adds a `summary_t`
tag then it should match the `*_t` dynamic field.  This is obvious and
requires no translation layer.  That handles the suffix issue, but
what about the prefix?  How does Yokozuna know _when_ a metadata key
is a tag?

A prefix like `tag-` could be used.  That's not so bad but it causes
metadata to be polluted with prefixes.  Also, what if the user wants
to tag keys like `content-type`, which already exist?  Should a
duplicate value be made with a tagged key?  Can the prefix be avoided
altogether?

Why not be explicit about which metadata keys are tags?  Add a special
metadata key named `tz-tags` containing a CSV of the metadata keys
to tag.  On the downside it requires an additional metadata entry but
on the upside it's explicit about what is being tagged, leaving the
user in full control.  It doesn't require remembering a set of rules
about when metadata is tagged.  This seems like the best possible
solution.
### Tags With CSV

In most cases it is obvious that the tag value should be sent to Solr
verbatim.  What about a comma separated value, CSV?  Solr has the
notion of multi-valued fields and it might make sense to treat a CSV
as such a field.  Once again the problem is discovering what semantic
the user wants.  A prefix of suffix could be used.  Perhaps a special
syntax for the `yz-tags` value, e.g. `yz-tags: keywords(csv)` would
indicate a CSV.  This could work but once again requires translation
and user education.  I think this should be avoided at all costs.

Instead, the tag value should be passed to Solr verbatim in all cases.
If a CSV should be converted into a multi-valued field then the user
should update the Solr config or schema to interpret it as such.

I think this could be achieved by either "poly fields" or a custom
[Field Mutating Update Processor](http://wiki.apache.org/solr/UpdateRequestProcessor).
## A Note on `X-` Headers

Notice the examples above are not using the `X-` convention for custom
HTTP headers.  The IETF deprecated this style in [RFC6648](http://tools.ietf.org/html/rfc6648) because
it causes problems for existing systems when a custom header becomes
part of the standard.  Also, using a prefix specific to your
application is sufficient.

A nice feature of the proposed plan above is that the user is in
complete control of the header names.  Either style may be used.

**Comment by coderoshi** [comment]
## Bucket/Index Tags

Considering the possible redundancy of setting `yz-tags` per-request, how about a configuration set as either a bucket property, or some other metadata added to the ring along with indexes.

This default set of keys could be set as part of the http index creation https://github.com/rzezeski/yokozuna/issues/8

Presumably, this tag list could optionally be overwritten per request, including the ability to suppress tagging with a blank header.

**Comment by rzezeski** [comment]
Tagging can't work as described above without changes to Riak KV.  To
store metadata in the Riak Object the
[headers must be prefixed](http://docs.basho.com/riak/latest/references/apis/http/HTTP-Store-Object/#Request) with `X-Riak-Meta`.  The
example above would change to the following.

```
X-Riak-Meta-yz-tags: X-Riak-Meta-myapp-where_loc, X-Riak-Meta-myapp-user_s, X-Riak-Meta-myapp-description_t
X-Riak-Meta-myapp-where_loc: Baltimore
X-Riak-Meta-myapp-user_s: rzezeski
X-Riak-Meta-myapp-description_t: Federal Hill at dusk.
```

Tagging will work with this new scheme but it feels clumsy.  Not to
mention potential overhead of disk for storing and CPU for processing.

A patch could be made to Riak KV to understand the `yz-tags` header.
Placing the listed tags in the object's metadata.  But this brings
knowledge of Yokozuna into KV.  I think it's worth trying to avoid
that, but perhaps the more pragmatic thing is to do just that.

Another approach is to strip the prefix off the tags before indexing
them in Solr.  E.g. `X-Riak-Meta-myapp-description` would be added to
the document as `myapp-description`.  The tags would still be stored
under the user metadata section of the object but the user searches on
the field without the prefix.  This keeps things clean on the query
side.

The kink with that plan is keeping parity with the protobuff
interface.  The user would have to know how to properly build a
user metadata entry in the object's metadata.  I.e. something that is
done automatically by the HTTP interface would have to be done
manually if using protobuff.

I think a pragmatic solution to this problem is to allow both the
user metadata format and the originally proposed format when using
protobuffs.  The clean `yz-tags` interface would work but if needed
the user metadata method could be used.

In summary, I see two options.
1. Modify Riak KV to understand `yz-tags`.  Store them in regular
   metadata, not special user metadata.
2. Translate HTTP user metadata to clean Solr fields by stripping
   prefix.  Allow either method for protobuff.  Requires no
   modification of Riak KV.

**Comment by rzezeski** [comment]
It seems that Riak Ojbect metadata removal may be lower than the HTTP
layer.  A [recent riak-users post](http://lists.basho.com/pipermail/riak-users_lists.basho.com/2012-October/009904.html) suggest so.  This means a
user of protobuff needs to manually create the user metadata entry for
tagging to work.  I think this is bogus.  I think a better approach
would be to add a header blacklist at the HTTP layer and allow
anything at the protbuff layer.  Or use the prefix method but strip
the prefix and store the entries in the top-level dictionary.  Keeping
all translation contained in the HTTP layer.  This would allow the
original proposal to be implemented.

But, being pragmatic for the 0.1 release I'm going with option 2 from
the last comment.  It's backwards compatible with the original
solution.  After the 0.1 release I can look at adding more sane user
metadata support to KV and update Yokozuna.

**Comment by rzezeski** [comment]
> Considering the possible redundancy of setting yz-tags per-request, how about a configuration set as either a bucket property, or some other metadata added to the ring along with indexes.

I'm going to punt on this.  It adds complexity both in terms of implementation and user understanding.  Also, in the majority of cases another program is writing to Riak thus the redundancy shouldn't matter.

**Comment by rzezeski** [comment]
Preliminary tagging support has been added for the 0.1 release.  I may open a new issue for 0.2 to see if it's possible to get rid of some of the `X-Riak-Meta` cruft.
